no-vowels/no-vowels.c

```diff
@@ -0,0 +1,52 @@
+// Write a function to replace vowels with numbers
+// Get practice with strings
+// Get practice with command line
+// Get practice with switch
+
+#include <cs50.h>
+#include <stdio.h>
+
+// Prototype
+void replace(char*);
+
+typedef struct {
+    char letter;
+    char symbol;
+} leet_map;
+
+leet_map map[8] = {
+    {'A', '6'},
+    {'a', '6'},
+    {'E', '3'},
+    {'e', '3'},
+    {'I', '1'},
+    {'i', '1'},
+    {'O', '0'},
+    {'o', '0'},
+};
+
+int main(int argc, string argv[])
+{
+    if (argc != 2)
+    {
+        printf("usage: ./no-vowels word\n");
+        return 1;
+    }
+    replace(argv[1]);
+}
+
+void replace(char* word)
+{
+    const int map_size = sizeof(map) / sizeof(map[0]);
+    for (int i = 0; word[i] != '\0'; i++)
+    {
+       for (int j = 0; j < map_size; j++)
+        {
+            if (word[i] == map[j].letter)
+            {
+                word[i] = map[j].symbol;
+            }
+        }
+    }
```

Comment:
It should be rare that you need to iterate an associative container - if you're having to iterate for a result, you should probably consider a different container. A key thing to consider here is that in many cases, iterating an associative container is functionally equivalent to iterating a linked list.

Let's consider the properties of an associative container. Insert/delete is O(1). Lookup is O(1). Iteration is O(n). In general, the best operation to use is the one with the smallest big O. So the question here is, can we reimplement this in terms of lookup instead of iteration. We can, and that implementation looks something like this:

```suggestion
    for (int i = 0; word[i] != '\0'; i++)
    {
       if (map.contains(word[i])) {
          word[i] = map[word[i]].symbol;
       }
    }
```

Considering the properties here, we've actually significantly improved things. For each letter in the input string, we now have just 1 operation instead of, on average, n/2 where n is the size of map!

Reply:
I understand; I knew there was something more powerful to do with a container but I forgot our `map` discussion from a few weeks ago.

Reply:
Is this pseudo-code?
```
       if (map.contains(word[i])) {
          word[i] = map[word[i]].symbol;
       }
```

I'm receiving the error:
```
error: member reference base type 'leet_map[8]' is not a structure or union
   43 |        if (map.contains(word[i])) 
      |            ~~~^~~~~~~~~
1 error generated.
```
and my poking around suggests `contains` doesn't exist in C. What have I missed?

Reply:
Yeah - that was just pseudocode. There should be some function for checking whether or not a map has a particular key.

Reply:
In order to use `contains` (or the C equivalent), I'd need to implement a hash function, search for the hashed value of the key in the map, then return the value of that key. Considering the need for an entire function, would I still gain any performance? (I understand the concept of n/2 but with no experience to calculate it.)

Reply:
It's a great question. There's not a universal answer here. In the case that you have 1 element, the overhead of the conversion to a hash vs just iterating through memory is obviously not worth it. In the case that you have 1 trillion elements the overhead of a hash calculation vs iterating (on-average) 500 billion elements seems like a pretty sweet trade off to make. Somewhere in the middle there the performance crosses over, but there's no universal point at which that happens - a lot of factors play in.

Since we're getting into the nitty gritty - for this application I suspect the fastest approach would be a [jump table](https://en.wikipedia.org/wiki/Branch_table). I think this is true because the conditions are a) simple and b) highly constrained (ie: compile time known and, in this case, few in number). I would tend to reach for a hash map when my problem had more runtime configurability (eg: if the user needed to be able to replace any character with another character, as specified on the command line).

The way to get a compiler to generate a jump table is either with a carefully crafted `if/else` or with a `switch`. In general, the preference is for `switch` because it's pretty much guaranteed to get the codegen that you want and it's simpler to write. In this case, that would look something like this:

```c
void replace(char* word) {
  for (int i = 0; word[i] != '\0'; ++i) {
    replace_char_if_necessary(word[i]);
  }
}

void replace_char_if_necessary(char* character) {
  switch (*character) {
    case 'a':
    case 'A':
      character = '6';
      break;
    case 'e':
    case 'E':
      character = '3';
      break;
    case 'i':
    case 'I':
      character = '1';
      break;
    case 'o':
    case 'O':
      character = '0';
      break;
  }
}
```

Reply:
Implemented and extra struct removed.